### PR TITLE
Add fov

### LINKED ISSUE #12: FOV
Fov in front that counts creatures in front and finds closest one

### PR DESCRIPTION
**Pull request by justinpaoli**
Fix #12

### CODE REVIEW

File: src/Helpers/PolarCoordHelper.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import {getDirection} from "./Vector2DHelper"
+
+export function convertPolarToCartesian(origin_x, origin_y, r, theta) {
```
Comment (devonpmack) [inline src/Helpers/PolarCoordHelper.js]: this should be a static function on the class

File: src/Helpers/PolarCoordHelper.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    this.y = origin_y;
+    this.r1 = r1;
+    this.r2 = r2;
+    this.theta1 = theta1;
+    this.theta2 = theta2;
+    this.span = this.theta2 > this.theta1 ? this.theta2 - this.theta1 : this.theta2 + 2 * Math.PI - this.theta1;
+  }
+
+  draw(p5) {
+    const line1_p1 = convertPolarToCartesian(this.x, this.y, this.r1, this.theta1);
```
Comment (devonpmack) [inline src/Helpers/PolarCoordHelper.js]: replace with 
```js
const line1 = {
p1: ...,
p2: ...
}
```

File: src/Helpers/PolarCoordHelper.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    p5.stroke(255);
+    p5.noFill();
+
+    p5.line(line1_p1.x, line1_p1.y, line1_p2.x, line1_p2.y);
+    p5.line(line2_p1.x, line2_p1.y, line2_p2.x, line2_p2.y);
+    p5.arc(this.x, this.y, 2 * this.r1, 2 * this.r1, this.theta1, this.theta2);
+    p5.arc(this.x, this.y, 2 * this.r2, 2 * this.r2, this.theta1, this.theta2)
+  }
+
+  // Working with angles is hell unless you do this trust me
```
Comment (devonpmack) [inline src/Helpers/PolarCoordHelper.js]: ok

File: src/Helpers/PolarCoordHelper.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  normalizeAngles() {
+    this.theta1 %= 2 * Math.PI;
+    this.theta2 %= 2 * Math.PI;
+  }
+
+  moveOrigin(x, y) {
+    this.x = x;
+    this.y = y;
+  }
```
Comment (devonpmack) [inline src/Helpers/PolarCoordHelper.js]: lol

File: src/Helpers/PolarCoordHelper.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    const delta_x = x - this.x;
+    const delta_y = y - this.y;
+    const r = Math.hypot(delta_x, delta_y);
+    const theta = getDirection(delta_x, delta_y);
+
+    if(this.theta2 > this.theta1) {
+      return this.r1 <= r && r <= this.r2 && this.theta1 <= theta && theta <= this.theta2;
+    } else {
+      return this.r1 <= r && r <= this.r2 && !(this.theta1 >= theta && theta >= this.theta2);
+    }
```
Comment (devonpmack) [inline src/Helpers/PolarCoordHelper.js]:
```js
return (this.theta2 > this.theta1) ? 
    this.r1 <= r && r <= this.r2 && this.theta1 <= theta && theta <= this.theta2 :
    this.r1 <= r && r <= this.r2 && !(this.theta1 >= theta && theta >= this.theta2);```

File: src/Helpers/Vector2DHelper.js [hunk 6]
```diff
@@ -0,0 +1,4 @@
+export function getDirection(dx, dy) {
+  const angle = Math.atan2(dy, dx);
+  return angle > 0 ? angle : angle + 2 * Math.PI;
+}
```
Comment (devonpmack) [inline src/Helpers/Vector2DHelper.js]: this is a sad file

**Comment by devonpmack** [comment]
If possible it'd be nice to make it a flag but not necessary.

File: src/Helpers/PolarCoordHelper.js (hunk 1 above)
Comment (justinpaoli) [reply src/Helpers/PolarCoordHelper.js]: It's not specific to polar rectangles, it's a general function to do with polar coordinates

File: src/Creature.js [hunk 7]
```diff
@@ -6,7 +6,8 @@ import {
   MUTATE_CHANCE,
   PREDATION,
   PREDATION_SIZE_MARKUP,
-  PREDATION_SPEED_MARKUP, SHOW_GENE_LABEL, USE_KINETIC_ENERGY
```
Comment (devonpmack) [inline src/Creature.js]: can we fix this newline

Comment (devonpmack) [reply src/Creature.js]: wtf

File: src/Creature.js [hunk 8]
```diff
@@ -37,7 +38,7 @@ export default class Creature {
     this.distance_remaining = Math.sqrt(genes.distance) * 150;
 
     this.sense = this.genes.sense * 25 + this.size / 2; // divide by 2 if accurate
-    this.fov = new PolarRect(this.x, this.y, 0, this.sense, 0, Math.PI / 2); // Initialize vision as a quarter circle
+    this.fov = USE_FOV ? new PolarRect(this.x, this.y, 0, this.sense, 0, Math.PI / 2): new PolarRect(this.x, this.y, 0, this.sense, 0, 2 * Math.PI); // Initialize vision as a quarter circle
```
Comment (devonpmack) [inline src/Creature.js]: what the fuck did you not listen to the creels advice of 40 characters on one line

File: src/Helpers/PolarCoordHelper.js [hunk 9]
```diff
@@ -16,24 +16,32 @@ export default class PolarRect {
...
-    const line2_p2 = convertPolarToCartesian(this.x, this.y, this.r2, this.theta2);
+    const line1 = {
+      p1: convertPolarToCartesian(this.x, this.y, this.r1, this.theta1),
+      p2: convertPolarToCartesian(this.x, this.y, this.r2, this.theta1)
+    };
+
+    const line2 = {
+      p1: convertPolarToCartesian(this.x, this.y, this.r1, this.theta2),
+      p2: convertPolarToCartesian(this.x, this.y, this.r2, this.theta2)
+    };
```
Comment (devonpmack) [inline src/Helpers/PolarCoordHelper.js]: very nice 🎉

File: src/Helpers/PolarCoordHelper.js [hunk 10]
```diff
@@ -42,17 +50,21 @@ export default class PolarRect {
...
 
-    this.normalizeAngles();
+      this.normalizeAngles();
+    }
   }
 
   rotate(angle) {
-    this.theta1 += angle;
-    this.theta2 += angle;
+    if (this.span < 2 * Math.PI) {
```
Comment (devonpmack) [inline src/Helpers/PolarCoordHelper.js]: idk why u added these everywhere but im sure its useful

File: src/Helpers/PolarCoordHelper.js [hunk 11]
```diff
@@ -61,10 +73,9 @@ export default class PolarRect {
...
 
-    if(this.theta2 > this.theta1) {
-      return this.r1 <= r && r <= this.r2 && this.theta1 <= theta && theta <= this.theta2;
-    } else {
-      return this.r1 <= r && r <= this.r2 && !(this.theta1 >= theta && theta >= this.theta2);
-    }
+    // Dear Shopify if you see this I used a multi-line ternary operator please hire me
+    return (this.theta2 > this.theta1) ? 
+      this.r1 <= r && r <= this.r2 && this.theta1 <= theta && theta <= this.theta2 :
+      this.r1 <= r && r <= this.r2 && !(this.theta1 >= theta && theta >= this.theta2);
```
Comment (devonpmack) [inline src/Helpers/PolarCoordHelper.js]: beautiful i will hire u for $1

**Pull request merged**